src/codegen.cpp

```diff
@@ -2321,6 +2325,52 @@ static Value *emit_known_call(jl_value_t *ff, jl_value_t **args, size_t nargs,
             }
         }
     }
+    else if (f->fptr == &jl_f_invoke && nargs >= 2 &&
+             expr_type(args[1], ctx) == (jl_value_t*)jl_function_type) {
+        // Get function
+        rt1 = static_eval(args[1], ctx);
+        f = (jl_function_t*)rt1;
+        if (!f || !jl_is_gf(f)) {
+            JL_GC_POP();
+            return NULL;
+        }
+        // Get types
+        jl_value_t *type_types = expr_type(args[2], ctx);
+        rt2 = type_types;
+        // Only accept Type{Tuple{...}}
+        if (!jl_is_type_type(type_types) ||
+            !jl_is_tuple_type(jl_tparam0(type_types))) {
+            JL_GC_POP();
+            return NULL;
+        }
+        rt2 = jl_tparam0(type_types);
+        // Get types of real arguments
+        jl_svec_t *aty = call_arg_types(&args[3], nargs - 2, ctx);
+        if (!aty) {
+            JL_GC_POP();
+            return NULL;
+        }
+        rt3 = (jl_value_t*)aty;
+        rt3 = (jl_value_t*)jl_apply_tuple_type(aty);
+        jl_function_t *mfunc =
+            jl_gf_invoke_get_specialization(f, (jl_tupletype_t*)rt2,
+                                            (jl_tupletype_t*)rt3);
+        JL_GC_POP();
+        if (mfunc == NULL) {
+            return NULL;
+        }
+        if (!is_constant(args[2], ctx))
+            emit_expr(args[2], ctx);
+        assert(mfunc->linfo->functionObject != NULL);
+        Value *_theF = literal_pointer_val((jl_value_t*)mfunc);
+        Value *_theFptr = emit_nthptr_recast(
```

Comment:
I thought the caller of this function assumes that the argument of the function call starts at 2?